repair/repair.cc

```diff
@@ -2052,6 +2052,77 @@ future<> repair_service::replace_with_repair(locator::token_metadata_ptr tmptr,
+    }
+    table_id tid = t.schema()->id();
+    // FIXME: we need to wait for current tablet movement and disable future tablet movment
+    auto erm = t.get_effective_replication_map();
+    auto& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+    struct repair_tablet_meta {
+        locator::tablet_id id;
+        dht::token_range range;
+        locator::host_id master_host_id;
+        shard_id master_shard_id;
+        locator::tablet_replica_set replicas;
+    };
+    std::vector<repair_tablet_meta> metas;
+    auto myhostid = erm->get_token_metadata_ptr()->get_my_id();
+    auto myip = erm->get_topology().my_address();
+    size_t total_tablets = 0;
+    size_t nr = 0;
+    co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) {
+        total_tablets++;
+        auto range = tmap.get_token_range(id);
+        auto& replicas = info.replicas;
+        bool found = false;
+        shard_id master_shard_id;
+        // Repair all tablets belong to this node
+        for (auto& r : replicas) {
+            if (r.host == myhostid) {
+                master_shard_id = r.shard;
+                found = true;
+                break;
+            }
+            if (primary_replica_only) {
+                break;
+            }
+        }
+        if (found) {
+            metas.push_back(repair_tablet_meta{id, range, myhostid, master_shard_id, replicas});
+        }
+    });
+    for (auto& m : metas) {
+        nr++;
+        rlogger.info("repair[{}] Repair {} out of {} tablets: table={}.{} tablet_id={} range={} replicas={} primary_replica_only={}",
+            rid.uuid(), nr, metas.size(), keyspace_name, table_name, m.id, m.range, m.replicas, primary_replica_only);
+        std::vector<gms::inet_address> nodes;
+        auto master_shard_id = m.master_shard_id;
+        auto range = m.range;
+        std::vector<shard_id> shards;
+        for (auto& r : m.replicas) {
+            auto shard = r.shard;
+            auto ip = co_await host2ip(r.host);
+            if (r.host != myhostid) {
+                rlogger.debug("repair[{}] Repair get neighbors table={}.{} hostid={} shard={} ip={} myip={} myhostid={}",
+                        rid.uuid(), keyspace_name, table_name, r.host, shard, ip, myip, myhostid);
+                shards.push_back(shard);
+                nodes.push_back(ip);
+            }
+        }
+        std::unordered_map<dht::token_range, repair_neighbors> neighbors;
+        neighbors[m.range] = repair_neighbors(nodes, shards);
+        co_await container().invoke_on(master_shard_id, [rid, keyspace_name, tid, neighbors, range] (repair_service& rs) -> future<> {
+            auto& t = rs._db.local().find_column_family(tid);
+            auto erm = t.get_effective_replication_map();
+            co_await rs.sync_tablet_using_repair(rid, keyspace_name, tid, erm, range, neighbors, streaming::stream_reason::repair);
+        });
```

Comment:
This repairs one tablet per node at a time, we should repair one tablet per shard.

Reply:
We could indeed split the work into multiple shards and start on per shard. One problem is that followers might have more than one tablet repair, e,g. tablet(n1:0 n2:1) tablet (n1:1 n2:1). Another things is that if we run nodetool repair -pr on all nodes in parallel, the node in question could run repair on other shards as well, other shards would be utilized even if the repair master work on one shard at a time.

Reply:
This problem will be solved when we implement the design based on tablet scheduler, which has global knowledge. If we want to do it earlier, we'd have to redo the scheduling code here. If it's not based on group0, then parallel nodetool invocations won't see each other.

I think we'd be better off investing in moving this to tablet scheduler.

Reply:
It will be a serious regression (slower by number of shards in a node). I think [this](https://github.com/scylladb/scylladb/pull/16600#discussion_r1440669935) is simple enough.